stackslib/src/net/api/callreadonly.rs

```diff
@@ -0,0 +1,390 @@
+                            // we want to execute any function as long as no actual writes are made as
+                            // opposed to be limited to purely calling `define-read-only` functions,
+                            // so use `read_only = false`.  This broadens the number of functions that
+                            // can be called, and also circumvents limitations on `define-read-only`
+                            // functions that can not use `contrac-call?`, even when calling other
+                            // read-only functions
+                            env.execute_contract(
+                                &contract_identifier,
+                                function.as_str(),
+                                &args,
+                                false,
+                            )
+                        },
+                    )
+                })
+            });
+
+        // decode the response
+        let data_resp = match data_resp {
+            Ok(Some(Ok(data))) => CallReadOnlyResponse {
+                okay: true,
+                result: Some(format!("0x{}", data.serialize_to_hex())),
+                cause: None,
+            },
+            Ok(Some(Err(e))) => match e {
+                Unchecked(CheckErrors::CostBalanceExceeded(actual_cost, _))
+                    if actual_cost.write_count > 0 =>
+                {
+                    CallReadOnlyResponse {
+                        okay: false,
+                        result: None,
+                        cause: Some("NotReadOnly".to_string()),
+                    }
+                }
+                _ => CallReadOnlyResponse {
+                    okay: false,
+                    result: None,
+                    cause: Some(e.to_string()),
+                },
+            },
+            Ok(None) | Err(_) => {
+                return StacksHttpResponse::new_error(
+                    &preamble,
+                    &HttpNotFound::new("Chain tip not found".to_string()),
+                )
+                .try_into_contents()
+                .map_err(NetError::from);
+            }
+        };
+
+        let mut preamble = HttpResponsePreamble::ok_json(&preamble);
+        preamble.set_canonical_stacks_tip_height(Some(node.canonical_stacks_tip_height()));
+        let body = HttpResponseContents::try_from_json(&data_resp)?;
+        Ok((preamble, body))
+    }
+}
+
+/// Decode the HTTP response
+impl HttpResponse for RPCCallReadOnlyRequestHandler {
+    fn try_parse_response(
+        &self,
+        preamble: &HttpResponsePreamble,
+        body: &[u8],
+    ) -> Result<HttpResponsePayload, Error> {
+        let map_entry: CallReadOnlyResponse = parse_json(preamble, body)?;
+        Ok(HttpResponsePayload::try_from_json(map_entry)?)
+    }
+}
+
+impl StacksHttpRequest {
+    /// Make a new request to run a read-only function
+    pub fn new_callreadonlyfunction(
+        host: PeerHost,
+        contract_addr: StacksAddress,
```

Comment:
I think you could make a lot of these arguments references since they are only ever used as references within this function. (e.g. contract_addr, contract_name, function_name)

Reply:
I try to use pass-by-value so the callee does not need to perform any extra allocations just to construct a request.  The caller almost always needs to perform them, however, but it does so solely to build a request.  I'm not sure pass-by-reference gains anything here?